### PR TITLE
grpc: Service for triggering indexer pruning

### LINKED ISSUE #1258: gRPC: Add endpoint to trigger indexer pruning
### Summary

Currently, there is no way for operators to trigger pruning of the indexer store. Following the logic implemented within ADR-101 (#82) in #1246 , we need to implement a way for a data companion to trigger indexer pruning.

The indexer pruning mechanism wil be implemented in #1176 .

### PR DESCRIPTION
**Pull request by werty144**
Closes #1258

---

#### PR checklist

- [x] Tests written/updated
- [x] Changelog entry added in `.changelog` (we use [unclog](https://github.com/informalsystems/unclog) to manage our changelog)
- [ ] Updated relevant documentation (`docs/` or `spec/`) and code comments

### CODE REVIEW

**Review by andynog — COMMENTED** [review]
added a few comments, overall looks good and tested the service in a client app

File: .changelog/unreleased/features/1260-grpc-pruning-indexer.md [hunk 1]
```diff
@@ -0,0 +1,2 @@
+- `[rpc/grpc]` Add grpc endpoint for pruning the tx_index.db.
```
Comment (thanethomson) [inline .changelog/unreleased/features/1260-grpc-pruning-indexer.md:1]:
```suggestion
- `[rpc/grpc]` Add gRPC endpoint for pruning the block and transaction indexes
```

File: .changelog/unreleased/improvements/169-tx_index-pruning.md [hunk 2]
```diff
@@ -0,0 +1,2 @@
+- `[state/indexer]` Add pruning for the tx_index.db.
```
Comment (thanethomson) [inline .changelog/unreleased/improvements/169-tx_index-pruning.md:1]:
```suggestion
- `[state/indexer]` Add background mechanism for pruning the block and transaction indexes
```

File: go.mod [hunk 3]
```diff
@@ -216,6 +216,7 @@ require (
 	github.com/nishanths/predeclared v0.2.2 // indirect
 	github.com/nunnatsa/ginkgolinter v0.13.3 // indirect
 	github.com/olekukonko/tablewriter v0.0.5 // indirect
+	github.com/onsi/ginkgo v1.16.4 // indirect
```
Comment (thanethomson) [inline go.mod:219]: Please run `go mod tidy`

File: state/pruner.go [hunk 4]
```diff
@@ -219,6 +230,58 @@ func (p *Pruner) SetABCIResRetainHeight(height int64) error {
 	return nil
 }
 
+func (p *Pruner) SetTxIndexerRetainHeight(height int64) error {
+	// Ensure that all requests to set retain heights via the application are
+	// serialized.
```
Comment (thanethomson) [inline state/pruner.go:235]:
```suggestion
	// Ensure that all requests to set retain heights are serialized.
```

File: state/pruner.go [hunk 5]
```diff
@@ -219,6 +230,58 @@ func (p *Pruner) SetABCIResRetainHeight(height int64) error {
...
+	if err := p.txIndexer.SetTxIndexerRetainHeight(height); err != nil {
+		return err
+	}
+	// TODO call metrics
+	return nil
+}
+
+func (p *Pruner) SetBlockIndexerRetainHeight(height int64) error {
+	// Ensure that all requests to set retain heights via the application are
+	// serialized.
```
Comment (thanethomson) [inline state/pruner.go:261]:
```suggestion
	// Ensure that all requests to set retain heights are serialized.
```

File: state/txindex/indexer_service.go [hunk 6]
```diff
@@ -33,7 +33,11 @@ func NewIndexerService(
 	terminateOnError bool,
 ) *IndexerService {
 
-	is := &IndexerService{txIdxr: txIdxr, blockIdxr: blockIdxr, eventBus: eventBus, terminateOnError: terminateOnError}
+	is := &IndexerService{
+		txIdxr:           txIdxr,
+		blockIdxr:        blockIdxr,
+		eventBus:         eventBus,
+		terminateOnError: terminateOnError}
```
Comment (thanethomson) [inline state/txindex/indexer_service.go:40]: Do you use a formatter locally in your IDE, like `gofumpt`?

**Review by thanethomson — COMMENTED** [review]
A quick first pass review - will do a deeper dive tomorrow!

File: rpc/grpc/client/privileged/pruning_service.go [hunk 7]
```diff
@@ -16,17 +16,59 @@ type RetainHeights struct {
 	PruningService uint64
 }
 
+type RetainHeight struct {
+	Height uint64
+}
+
```
Comment (mzabaluev) [inline rpc/grpc/client/privileged/pruning_service.go:22]: I would not invent a struct type for just this one field, just return a `uint64` like in the other getters returning a scalar.
It's also awfully ambiguous with `RetainHeights` returned by `GetBlockRetainHeight` (which, I admit, was also ambiguously named).

File: rpc/grpc/client/privileged/pruning_service.go [hunk 8]
```diff
@@ -110,15 +110,15 @@ func (c *disabledPruningServiceClient) SetTxIndexerRetainHeight(_ context.Contex
...
+func (c *disabledPruningServiceClient) GetTxIndexerRetainHeight(_ context.Context) (uint64, error) {
 	panic("pruning service client is disabled")
 }
 
 func (c *disabledPruningServiceClient) SetBlockIndexerRetainHeight(_ context.Context, _ uint64) error {
 	panic("pruning service client is disabled")
 }
 
-func (c *disabledPruningServiceClient) GetBlockIndexerRetainHeight(ctx context.Context) (uint64, error) {
+func (c *disabledPruningServiceClient) GetBlockIndexerRetainHeight(_ context.Context) (uint64, error) {
```
Comment (jmalicevic) [inline rpc/grpc/client/privileged/pruning_service.go:121]: Why are you passing a context that you are not using?

File: state/store_test.go [hunk 9]
```diff
@@ -314,17 +314,7 @@ func TestSaveRetainHeight(t *testing.T) {
...
-	require.Error(t, err)
-
-	// We should not save a height above the blockstore's height
-	err = pruner.SetApplicationBlockRetainHeight(11)
-	require.Error(t, err)
-
-	err = pruner.SetApplicationBlockRetainHeight(10)
-
-	err = initStateStoreRetainHeights(stateStore, 0, 0, 0)
+	err := initStateStoreRetainHeights(stateStore, 0, 0, 0)
```
Comment (jmalicevic) [inline state/store_test.go:317]: Why is this part of this PR?

File: proto/tendermint/services/pruning/v1/service_grpc.pb.go [hunk 10]
```diff
@@ -61,7 +61,7 @@ func NewPruningServiceClient(cc grpc.ClientConnInterface) PruningServiceClient {
 
 func (c *pruningServiceClient) SetBlockRetainHeight(ctx context.Context, in *SetBlockRetainHeightRequest, opts ...grpc.CallOption) (*SetBlockRetainHeightResponse, error) {
 	out := new(SetBlockRetainHeightResponse)
-	err := c.cc.Invoke(ctx, "/v1.PruningService/SetBlockRetainHeight", in, out, opts...)
+	err := c.cc.Invoke(ctx, "/tendermint.services.pruning.v1.PruningService/SetBlockRetainHeight", in, out, opts...)
```
Comment (jmalicevic) [inline proto/tendermint/services/pruning/v1/service_grpc.pb.go:64]: This should not have been changed.

File: proto/tendermint/services/pruning/v1/service_grpc.pb.go [hunk 11]
```diff
@@ -70,7 +70,7 @@ func (c *pruningServiceClient) SetBlockRetainHeight(ctx context.Context, in *Set
 
 func (c *pruningServiceClient) GetBlockRetainHeight(ctx context.Context, in *GetBlockRetainHeightRequest, opts ...grpc.CallOption) (*GetBlockRetainHeightResponse, error) {
 	out := new(GetBlockRetainHeightResponse)
-	err := c.cc.Invoke(ctx, "/v1.PruningService/GetBlockRetainHeight", in, out, opts...)
+	err := c.cc.Invoke(ctx, "/tendermint.services.pruning.v1.PruningService/GetBlockRetainHeight", in, out, opts...)
```
Comment (jmalicevic) [inline proto/tendermint/services/pruning/v1/service_grpc.pb.go:73]: Same as above.

File: proto/tendermint/services/pruning/v1/service_grpc.pb.go [hunk 12]
```diff
@@ -394,5 +394,5 @@ var PruningService_ServiceDesc = grpc.ServiceDesc{
 		},
 	},
 	Streams:  []grpc.StreamDesc{},
-	Metadata: "proto/tendermint/services/pruning/v1/service.proto",
+	Metadata: "tendermint/services/pruning/v1/service.proto",
```
Comment (jmalicevic) [inline proto/tendermint/services/pruning/v1/service_grpc.pb.go:397]: Should not happen

File: go.mod (hunk 3 above)
Comment (andynog) [reply go.mod]: fixed

File: state/txindex/indexer_service.go (hunk 6 above)
Comment (andynog) [reply state/txindex/indexer_service.go]: fixed

File: rpc/grpc/client/privileged/pruning_service.go (hunk 7 above)
Comment (andynog) [reply rpc/grpc/client/privileged/pruning_service.go]: fixed

**Comment by werty144** [comment]
Superseded by #1327

**Pull request closed without merging**